Make it so that sympy.codegen is not imported with 'from sympy import *'

asmeurer opened the pull request:
This moves the ccode, cxxcode, and fcode top-level printers to the
codeprinter.py file (they are also re-imported in the original files for
backwards compatibility).

This PR has two minor compatibility breaks.

- 'codegen' is no longer imported with 'from sympy import *'. However, all
  top-level module names being imported like this is deprecated, so this one
  is just an early removal.

- 'from sympy.codegen.codeprinter import Assignment' no longer works. This was
  kept intact for backwards compatibility. Assignment should now be imported
  from sympy.codegen.ast, or just sympy.codegen, which is where it has lived
  for several releases now.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

This is part of my work at #19907.

#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- codegen
  - `sympy.codegen` and `sympy.combinatorics` are no longer imported as part of `from sympy import *`. The codegen and combinatorics modules do not have any names that are included in the top-level namespace, so their being imported with the top-level SymPy was a mistake. This improves the import time of `from sympy import *`.
- printing
  - **BACKWARDS COMPATIBILITY BREAK:** The modules `sympy.printing.ccode`, `sympy.printing.fcode`, and `sympy.printing.cxxcode` have been renamed to `sympy.printing.c`, `sympy.printing.fortran`, and `sympy.printing.cxx`, respectively. This was done to avoid conflicts that occur when a module name is the same as a function name. 
  - `from sympy.printing.codeprinter import Assignment` no longer works. You should use `from sympy.codegen import Assignment`, which is where `Assignment` has lived since SymPy 1.1.

<!-- END RELEASE NOTES -->

asmeurer committed:
Make it so that sympy.codegen is not imported with 'from sympy import *'

This moves the ccode, cxxcode, and fcode top-level printers to the
codeprinter.py file (they are also re-imported in the original files for
backwards compatibility).

This PR has two minor compatibility breaks.

- 'codegen' is no longer imported with 'from sympy import *'. However, all
  top-level module names being imported like this is deprecated, so this one
  is just an early removal.

- 'from sympy.codegen.codeprinter import Assignment' no longer works. This was
  kept intact for backwards compatibility. Assignment should now be imported
  from sympy.codegen.ast, or just sympy.codegen, which is where it has lived
  for several releases now.

asmeurer committed:
Add a test that only expected submodules are imported with 'from sympy import *'

This test runs as a separate script because it has to import sympy before it
has been imported yet.

asmeurer commented:
I also added a test to prevent this from happening in the future. We might want to make the test actually check against `sympy/__init__.py`, but for now, the whitelist is manual.

asmeurer committed:
Fix code quality check

oscarbenjamin commented on bin/test_external_imports.py:
```diff
@@ -1,10 +1,10 @@
-#!/usr/bin/env python
+o#!/usr/bin/env python
```
I guess this `o` isn't intended

asmeurer replied:
Oops. emacs typo

asmeurer committed:
Fix a typo

oscarbenjamin commented:
The tests have failed with `TypeError: 'module' object is not callable`.

I think that's because of a conflict between whether
```
from sympy.printing import ccode
```
should give the function or the module.

asmeurer commented:
Odd, if I run that in both master and this branch, I get the function, and when I run `./bin/doctest sympy/tensor/indexed.py`, it passes. It only fails when I run the whole doctest suite.

asmeurer commented:
Oh I see, if you run `import sympy.printing.fcode` it causes `from sympy.printing import fcode` to give the module. It can be fixed by adding

```py
from . import fcode
from . import ccode
```

above

```py
from .codeprinter import ccode, print_ccode, cxxcode, fcode, print_fcode
```

in `sympy/printing/__init__.py`. But I wonder if there is another way to fix it.

asmeurer commented:
I guess the other "fix" is to not import those names from `sympy.printing` but to only import them from the top-level. The advantage is not having to import those two files at import time. But it is a compatibility break for anyone who is doing that. So for now I will go with the above fix. I have to say, the way Python importing works where it just injects names into the module namespace is a huge mess. Though it's also apparently a bad idea to ever have a module and function that share the same name, even though that is quite often a very natural way to name things.

asmeurer committed:
Fix 'from sympy.printing import ccode' after running 'import sympy.printing.ccode'

If ccode.py is not yet imported, then sympy.printing.ccode will be the
function, as imported in sympy/printing/__init__.py from
sympy/printing/codeprinter.py. But once sympy.printing.ccode is imported,
Python replaces it with the submodule, which breaks the above import.

To fix this, we have to make sure that the submodules ccode.py and fcode.py
are imported before the functions in sympy/printing/__init__.py. This is
somewhat unfortunate, as we would otherwise not even need to import those
submodules at all at top-level import time.

The only other alternative way to fix this that I can think of is to disallow
'from sympy.printing import ccode' and require 'from sympy import ccode'.
However, the former form already appears in several docstring examples, and it
is generally the sort of thing that people do in actual code, so it would be
very confusing to let this behavior happen.

asmeurer committed:
Also fix sympy.printing.cxxcode

See the previous commit.

asmeurer committed:
Organize some code lines a little better

eric-wieser commented on sympy/printing/ccode.py:
```diff
@@ -733,141 +736,3 @@ def _print_alignof(self, expr):
 }
-
-
-def ccode(expr, assign_to=None, standard='c99', **settings):
```
It seems weird to move this function - `latex` and `print_latex` are within `latex.py`, so keeping the convenience functions next to the printer subclasses looks to be the norm.

eric-wieser replied:
I suppose the change here is that the convenience functions no longer import their dependencies until needed for the first time, which I guess improved import time.

Can we make all the printers consistent in this approach? Or can we make users just write the longer import to explicitly import `print_fcode` etc?

asmeurer replied:
Yes, specifically the sympy.codegen module does not need to be imported at all when sympy is imported. Unfortunately, the codegen objects are used in the class body, so there's no way to avoid importing them at the top-level of this file. So the next best thing is to just avoid importing this file. 

In general I agree that this isn't a great change in terms of code organization, but I think it's worth it for this case because it removes an entire otherwise optional submodule from the top-level import. 

> Can we make all the printers consistent in this approach?

I was originally going to move all the printers around like this, but I decided against it. It's better to have the printers in the same files if possible. It's just these three printers that make extensive use of the codegen submodule. 

> Or can we make users just write the longer import to explicitly import print_fcode etc?

Well if users use the shorter imports like `from sympy import print_fcode` or even `from sympy.printing import print_fcode`, then it won't matter what file `print_fcode` is in. So I actually think that users importing things from fully qualified modules is not good, because it limits how we can reorganize code.

asmeurer replied:
And in general, Python's import semantics are often at odds with good code organization. You often find yourself fighting them to get the organization you want, either because of circular import issues, or issues like this where you want to avoid importing a module you don't need to, or other things. Sometimes you can make some compromises and win, e.g., by moving imports inside of functions. Other times you just have to give in and accept that your code can't be organized perfectly. I'm not aware of a general solution that always lets you organize your code in the way that is most natural, and also avoid issues with imports.

eric-wieser replied:
How viable is moving imports inside functions here?

asmeurer replied:
The objects are used at the class definition level https://github.com/sympy/sympy/pull/19908/files/01563a4d26ae825c909a4668bd5fdba9685ed618#diff-7b342526581a07bce690949be8806f6bR164-R197. So it would not be very easy without changing how the class works.

eric-wieser replied:
That link anchor doesn't go anyway - can you link to the lines on master / some git sha?

asmeurer replied:
https://github.com/sympy/sympy/blob/8544f01468187b9b27d4fdb4671ce672d03622f9/sympy/printing/ccode.py#L161-L213

oscarbenjamin commented:
I think we should stop having modules that have the same name as the functions in the modules

asmeurer commented on sympy/printing/__init__.py:
```diff
@@ -10,13 +10,18 @@
+# We have to import these now because otherwise if there are imported later,
+# they will override sympy.printing.ccode and so on with the
+# modules. It would be nice to not have to do this as we could avoid importing
+# these files at top-level import time completely otherwise.
```
Ah, of course we can't do this, because this makes sympy.codegen be imported again, which defeats the whole purpose. So I think the only fix here is to rename the files.

asmeurer committed:
Rename ccode.py, fcode.py, and cxxcode.py to c.py, fortran.py, and cxxcode.py

This is unfortunately a backwards incompatible change for anyone who imports
the printer classes from these submodules. However, it the only way to avoid
importing sympy.codegen at the top-level. The reason is this:

- If the module names share the same name as the function names, like ccode.py
  and ccode(), then whichever is imported last will be set as
  sympy.printing.ccode.

- sympy/printing/ccode.py (now sympy/printing/c.py) imports several things
  from sympy.codegen at the top-level. These are used in the class definition,
  so they are not easy to remove. Thus, if it is imported at the top-level,
  then so will sympy.codegen. We want to avoid this, as sympy.codegen should
  be a completely optional submodule and importing it at the top-level import
  time just adds to the import time.

- As a result, we have moved, the functions ccode(), print_ccode(), and so on
  to sympy.printing.codeprinter, so that they can be imported without
  importing sympy.printing.ccode.  However, this means that at first import,
  sympy.printing.ccode will be the function. But then as soon as the module
  sympy.printing.ccode is imported, it will replace this with the module,
  meaning that 'from sympy.printing import ccode' will start returning the
  module instead of the function.

There are only two ways to avoid this behavior. One is to make sure that
sympy.printing.ccode is imported before the ccode() function, as whichever one
is imported last will be sympy.printing.ccode. However, this requires
importing sympy.codegen at the top-level. The other way is to rename the
module, so that it doesn't conflict with the function name.

Making module names not conflict with function names is a good idea in
general, as it avoids this problem. However, it does mean a backwards
comparability break for any existing module. I don't know how we should
approach this for other such modules in SymPy. Currently, only these are
changed, as it is required to avoid the top-level sympy.codegen import.

asmeurer committed:
Print a success message in test_submodule_imports.py

asmeurer commented:
OK, I've changed this so that the modules `sympy.printing.ccode`, `sympy.printing.fcode`, and `sympy.printing.cxxcode` have been renamed to `sympy.printing.c`, `sympy.printing.fortran`, and `sympy.printing.cxx`. Unfortunately, it's a minor backwards compatibility break, but this is the only way I know of to avoid the module name from overwriting the function name, and not importing them at the top-level (which would mean importing `sympy.codegen`). I suppose we could also play around with lazy module objects. But as @oscarbenjamin said, we should avoid naming modules the same names as functions, so this is a good change to make, even if fixing it does mean breaking compatibility.

asmeurer committed:
Merge branch 'master' into no-codegen-import

oscarbenjamin commented:
This looks good to me.

Does the whitelist include everything except physics and stats?

I notice that combinatorics is in the whitelist and is imported by import sympy even though the main `__init__.py` deliberately doesn't import combinatorics apparently on the grounds that it is slow. At some point I made it faster to import combinatorics by precomputing some of structures. We should decide whether it should be imported or not.

asmeurer commented:
I generated the whitelist based on what was already being imported. I didn't notice that `combinatorics` is being imported only from the deprecated list. I can remove it.

asmeurer committed:
Don't import combinatorics with 'import sympy'

It was only imported because of the behavior that kept it as an deprecated
imported module with 'from sympy import *'.

asmeurer commented:
According to the `bin/test_import` script, in master, `import sympy` takes  0.314030 +- 0.023456 seconds, and in this branch it takes 0.273625 +- 0.006257 seconds.

asmeurer commented:
These are the submodules not imported with `import sympy`:

```
abc.py
benchmarks
categories
codegen
combinatorics
conftest.py
crypto
diffgeom
galgebra.py
holonomic
liealgebras
operator
physics
release.py
sandbox
stats
this.py
unify
vector
```

oscarbenjamin commented:
Looks good
